refactor: switch to interpreter (#218)
asyncapi/generator-model-sdk: docs/interpretation_of_JSON_Schema_draft_7.md

The algorithm tries to get to a model whose data can be validated against the JSON schema document. 

> We only provide the underlying structure of the schema file for the model, where formats such as `maxItems`, `uniqueItems`, `multipleOf`, etc, are not transformed.

Reviewer: ```suggestion
We only provide the underlying structure of the schema file for the model, where constraints/annotations such as `maxItems`, `uniqueItems`, `multipleOf`, etc, are not transformed.
```

- boolean `not` schemas are not applied.

## allOf sub schemas
`allOf` is a bit different than the other [combination keywords](#Processing-sub-schemas) since it can imply inheritance. 

So dependant on whether the interpreter option `allowInheritance` is true or false we interpret it as inheritance or [merge](#Merging-models) the models.

## Processing sub schemas
The following JSON Schema keywords are [merged](#Merging-models) with the already interpreted model:
- `oneOf`
- `anyOf`
- `then`
- `else`

## Merging models

> Because of the recursive nature of the interpreter (and the nested behavior of JSON Schema) it happens that two models needs to be merged together.

Reviewer: ```suggestion
Because of the recursive nature of the interpreter (and the nested nature of JSON Schema) it happens that two models needs to be merged together.
```

If only one side has a property defined, it is used as is, if both have it defined they are merged based on the following logic:

> - `additionalProperties` if both models contain it the two are recursively merged together. 

Reviewer: ```suggestion
- `additionalProperties` if both models contain items they are recursively merged together.
```

Author: Where does `items` fit in? 🤔

Reviewer: Sorry 😅 My fault. I did a review right after waking up. It can stay as is, it makes more sense :)

- `patternProperties` if both models contain a pattern the corresponding models are recursively merged together. 
- `properties` if both models contain the same property the corresponding models are recursively merged together. 
- `items` if both models contain items they are recursively merged together.
- `types` if both models contain types they are merged together, duplicate types are removed.
- `enum` if both models contain enums they are merged together, duplicate enums are removed.
- `required` if both models contain required properties they are merged together, duplicate required properties are removed.
- `$id`, `$ref`, `extend` uses left model value if present otherwise right model value if present.

> - `originalSchema` are overwritten.

Reviewer: We have information about `originalSchema` only in `API.md` file, we should described what is `originalSchema` here in another place :)

Author: Hmm, yea true. What do you think is the best solution? 🤔

Reviewer: It can be described here, or somewhere at the beginning of this doc :)

Author: Added a link to the input processor docs which explains the CommonModel, otherwise, we have to keep in mind that we have a description of a property in this document as well.
